subsys/net/lib/prometheus/formatter.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2024 Mustafa Abdullah Kus, Sparse Technology
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/prometheus/formatter.h>
+#include <zephyr/net/prometheus/collector.h>
+#include <zephyr/net/prometheus/metric.h>
+#include <zephyr/net/prometheus/histogram.h>
+#include <zephyr/net/prometheus/summary.h>
+#include <zephyr/net/prometheus/gauge.h>
+#include <zephyr/net/prometheus/counter.h>
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <stddef.h>
+
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pm_formatter, CONFIG_PROMETHEUS_LOG_LEVEL);
+
+static int write_to_buffer(char *buffer, size_t buffer_size, const char *str)
+{
+	/* helper function to write formatted string to buffer, handling buffer size */
+	int len = snprintf(buffer + strlen(buffer), buffer_size - strlen(buffer), "%s", str);
+
+	if (len >= 0 && len < buffer_size - strlen(buffer)) {
+		return 1;
+	}
```

Comment:
Zephyr functions typically return a negative error code with a value from `errno.h`.
For instance, `-ENOMEM` or `-ENAMETOOLONG`.